### PR TITLE
Implement XACKDEL and XDELEX command

### PR DESCRIPTION
**Pull request by nickiaq**
This commit implements `XACKDEL` and `XDELEX`. These commands make it easier to delete messages from a stream once all consumer groups have acknowledged the message or can no longer receive it.

For more background on the use case for these commands, see [this issue](https://github.com/valkey-io/valkey/issues/2903) or [this article](https://blog.arcjet.com/replacing-kafka-with-redis-streams/). Prior to these commands (since replicating this behavior is non-trivial in Lua), applications would run a separate "janitor" process that tracks each consumer group's progress and periodically calls XTRIM to reclaim space once all messages up to some high water mark have been ack'd by all groups.

For full Redis compatibility, the command supports the same three deletion modes with the same semantics, command syntax, and reply format:

- `KEEPREF` (default, implicit) deletes the message, leaving PEL references in other groups
    - For `XACKDEL`, this also acknowledges the message
- `ACKED` only deletes once every consumer group has acknowledged or can no longer receive the message, making it safe for fan-out stream topologies
    - In `XACKDEL`, if the message was ack'd but not deleted, the response is 2
    - In `XDELEX`, if there are still consumer groups that can receive the message (or have it in pending) the response is 2
- `DELREF` acks, deletes, and forcibly removes PEL entries from all other groups

`XACKDEL` and `XDELEX` return an array of status codes, one for each ID passed in. There are 3 possible values:

- `-1`: the key is not a stream, stream message does not exist
    - For `XACKDEL` this also covers when the group does not exist or the message was not in the target group's PEL
- `1`: the message was deleted; for `XACKDEL` this also indicates the message was acknowledged
- `2`: the message was acknowledged (for `XACKDEL`) but not deleted because other groups have it pending or could still receive it
    - There are some nuanced edge cases here around `XGROUP SETID` and/or the message being deleted vs in PEL

To ensure backwards compatibility, these commands are replicated as the equivalent `XACK`, `XDEL` commands so that a pre-9.2 replica can achieve the same effect that the `XACKDEL`, `XDELEX` on the primary had.

There are tests covering each mode of each command and many edge cases.

### CODE REVIEW

File: tests/unit/type/stream.tcl [hunk 1]
```diff
@@ -634,6 +634,762 @@ start_server {
...
+        set ids [r XDELEX teststream ACKED IDS 1 1-0]
+        assert_equal 2 [lindex $ids 0]
+        assert_equal 1 [r XLEN teststream]
+
+        # grp2 acks; all groups done → deletion occurs
+        r XACK teststream grp2 1-0
+        set ids [r XDELEX teststream ACKED IDS 1 1-0]
+        assert_equal 1 [lindex $ids 0]
+        assert_equal 0 [r XLEN teststream]
+    }
```
[the remark this answers is not in the document]
Comment (nickiaq) [reply tests/unit/type/stream.tcl:1491]: Addressed in 45effc1

File: src/t_stream.c [hunk 2]
```diff
@@ -3569,48 +3626,433 @@ void xdelCommand(client *c) {
...
+    }
+
+    *ids_argi = argi;
+    return C_OK;
+}
+
+/* XDELEX <key> [KEEPREF | DELREF | ACKED] IDS num [<ID1> <ID2> ... <IDN>]
+ */
+void xdelexCommand(client *c) {
+    robj *o = lookupKeyRead(c->db, c->argv[1]);
```
[the remark this answers is not in the document]
Comment (nickiaq) [reply src/t_stream.c]: Addressed in 45effc1

File: src/commands/xdelex.json [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            {
+                "name": "numids",
+                "type": "integer"
+            },
+            {
+                "name": "id",
+                "token": "IDS",
+                "type": "string",
+                "multiple": true
+            }
```
[the remark this answers is not in the document]
Comment (nickiaq) [reply src/commands/xdelex.json:79]: Addressed in 45effc1

**Comment by nickiaq** [comment]
I will submit fixes tomorrow morning for the coderabbit & valkey review bot comments. These are both valid.

For the coderabbit one, XACKDEL already handles this due to the 2 phase setup there. But for XDELEX, we have to track existance as it suggests.

For the issue replicating new commands to older replicas, I'll use `preventCommandPropagation` and instead emit the equivalent of the new behavior using pre-9.2 commands `XACK`, `XDEL`, etc.

**Comment by hpatro** [comment]
> For the issue replicating new commands to older replicas, I'll use `preventCommandPropagation` and instead emit the equivalent of the new behavior using pre-9.2 commands `XACK`, `XDEL`, etc.

This was a nice catch.

File: src/t_stream.c [hunk 4]
```diff
@@ -3569,48 +3626,433 @@ void xdelCommand(client *c) {
...
+}
+
+typedef enum {
+    PELMODE_KEEPREF = 0,
+    PELMODE_DELREF,
+    PELMODE_ACKED
+} streamPELMode;
+
+/* Parse the "[KEEPREF | DELREF | ACKED] IDS <numids>" arguments shared by
+ * XDELEX and XACKDEL. */
```
Comment (zuiderkwast) [inline src/t_stream.c]:
```suggestion
 * XDELEX and XACKDEL. Returns C_OK, or replies to the client and returns
 * C_ERR. */
```

File: src/t_stream.c [hunk 5]
```diff
@@ -3551,6 +3551,63 @@ void xautoclaimCommand(client *c) {
     preventCommandPropagation(c);
 }
 
+/* Parse IDS <numids> into array of stream message ids, ensuring each is a valid
+ * stream message ID. Returns C_OK, or replies to the client on first invalid
+ * ID and returns C_ERR. */
+static int streamParseDelIDsOrReply(client *c, int argi, long long id_count, streamID *ids, int *resps) {
```
Comment (zuiderkwast) [inline src/t_stream.c]: Let's name it `streamParseStrictIDsOrReply` (plural of the existing `streamParseStrictIDOrReply`) and move it next to that function definition, around line 1958.

**Review by zuiderkwast — COMMENTED** [review]
Thanks for combining the two PRs and avoiding some code duplication.

Pulling common code to helper functions is a good approach, but there is still some duplication in xackdelCommand / xdelexCommand, and the helper functions streamDeleteItemAndTrackFirstLast and streamTrackFirstEntryAndPropagate are slightly unintuitive, pieces of code pulled out of their context.

I have another idea, which is used in many places for related commands: Use a common function to implement all three commands. Idea:

```C
/* Shared implementation of XDEL/DELEX/ACKDEL commands.
 *
 * XDEL key [id...]
 * XDELEX key [pelmode] IDS numids [id...]
 * XACKDEL key group [pelmode] IDS numids [id...]
 */
void xdelGenericCommand(client *c, bool has_group_arg, bool has_pelmode_arg) {
    streamCG *group = NULL;
    streamPELMode mode = PELMODE_KEEPREF;
    robj *o = lookupKeyWrite(c->db, c->argv[1]);
    int argi = 2;

    if (o && checkType(c, o, OBJ_STREAM)) return; /* Type error. */

    if (o && has_group_arg) {
        group = streamLookupCG(objectGetVal(o), objectGetVal(c->argv[argi]));
        argi++;
    }

    if (has_pelmode_arg) {
        if (strcasecmp(objectGetVal(c->argv[argi]), "KEEPREF") == 0) {
            argi += 1;
        } else if (strcasecmp(objectGetVal(c->argv[argi]), "DELREF") == 0) {
            argi += 1;
            mode = PELMODE_DELREF;
        } else if (strcasecmp(objectGetVal(c->argv[argi]), "ACKED") == 0) {
            argi += 1;
            mode = PELMODE_ACKED;
        }

        /* Expect IDS token. */
        if (strcasecmp(objectGetVal(c->argv[argi]), "IDS") != 0) {
            addReplyErrorObject(c, shared.syntaxerr);
            return;
        }
        argi++; /* past IDS */

        /* Parse and validate numids: must be a positive integer. */
        long long id_count = 0;
        if (getLongLongFromObject(c->argv[argi], &id_count) != C_OK || id_count <= 0) {
            addReplyError(c, "Number of IDs must be a positive integer");
            return;
        }
        argi++; /* past numids */

        /* Validate numids matches remaining arg count. */
        if (id_count != c->argc - argi) {
            addReplyErrorObject(c, shared.syntaxerr);
            return;
        }
    }

    ...
}

void xdelCommand(client *c) {
    xdelGenericCommand(c, false, false);
}
void xdelexCommand(client *c) {
    xdelGenericCommand(c, false, true);
}
void xackdelCommand(client *c) {
    xdelGenericCommand(c, true, true);
}
```

In this way, all three are basically the same command. The code can be read from top to bottom. Would it work?

File: src/t_stream.c [hunk 6]
```diff
@@ -3569,48 +3626,433 @@ void xdelCommand(client *c) {
...
+                } else if (mode == PELMODE_ACKED) {
+                    if (first_loop && !streamEntryExists(s, id)) {
+                        /* Message doesn't exist in the stream; check once and
+                         * skip iterating the remaining groups. */
+                        resps[j] = -1;
+                    } else if (streamCompareID(id, &cg->last_id) > 0) {
+                        /* Message may still be delivered to this group, so
+                         * block deletion (same as XACKDEL). */
+                        resps[j] = 2;
+                    }
```
[the remark this answers is not in the document]
Comment (nickiaq) [reply src/t_stream.c]: Addressed in 71e9dbd2b.

File: src/t_stream.c [hunk 7]
```diff
@@ -3569,48 +3626,433 @@ void xdelCommand(client *c) {
...
-    if (deleted) {
-        if (s->length == 0) {
-            s->first_id.ms = 0;
-            s->first_id.seq = 0;
-        } else if (first_entry) {
-            streamGetEdgeID(s, 1, 1, &s->first_id);
-        }
+sync:
+    /* Stream bookkeeping. */
+    streamTrackFirstEntryAndPropagate(c, s, deleted, acked, first_entry);
```
[the remark this answers is not in the document]
Comment (nickiaq) [reply src/t_stream.c]: Addressed in 295a4e1ef.

File: tests/support/server.tcl [hunk 8]
```diff
@@ -618,7 +618,11 @@ proc start_server {options {code undefined}} {
     }
 
     set unixsocket [file normalize [format "%s/%s" [dict get $config "dir"] "socket"]]
-    dict set config "unixsocket" $unixsocket
+    # Skip unix socket if path is too long. macOS/BSD sun_path limit is 104
+    # & linux sun_path limit is 108.
+    if {[string length $unixsocket] < 104} {
+        dict set config "unixsocket" $unixsocket
+    }
```
[the remark this answers is not in the document]
Comment (nickiaq) [reply tests/support/server.tcl]: Sorry, I was planning to put this up as a separate PR and accidentally committed here. I've reverted this part in 9fdb69aec.

The issue here is that on MacOS/BSD the max socket path is 104. On Linux it's 107 chars. If the working directory of your checkout is more than this long, the unit tests fail.

For example:

```
mkdir -p /tmp/shallow
git clone https://github.com/valkey-io/valkey /tmp/shallow/valkey
cd /tmp/shallow/valkey
make -j $(nproc) test  # works

mkdir -p /tmp/this/is/a/very/deep/that/just/keeps/going/and/going/to/a/directory/that/is/over/the/socket/path/size/limit
git clone https://github.com/valkey-io/valkey /tmp/this/is/a/very/deep/that/just/keeps/going/and/going/to/a/directory/that/is/over/the/socket/path/size/limit/valkey
cd /tmp/this/is/a/very/deep/that/just/keeps/going/and/going/to/a/directory/that/is/over/the/socket/path/size/limit/valkey
make -j $(nproc) test  # hangs b/c server can't start due to bad socket path
```

Due to the length of the socket path part itself (~30 chars), this happens with checkout base directory paths as small as 74 characters.

The other solution to this is to make the socket path relative, but that was also hanging and I haven't had a chance to chase that approach down further yet.

**Comment by nickiaq** [comment]
Thanks @zuiderkwast and @hpatro for continued comments. To summarize the changes since last night, I've got corrections up for both the bot comments:

- 295a4e1ef: Addresses the Valkey Review Bot comments about backwards compatibility, switching propagation to use `XDEL`, `XACK` to replicate the equivalent effect of the `XACKDEL` and `XDELEX` command's mutations. (Also adds tests to `stream-cgroups.tcl` checking what's replicated)
- 71e9dbd2b: Addressing the CodeRabbit comments about improperly handling dangling PEL entries for previously deleted stream messages (and adding tests covering this)

And towards removing duplicated code, I have added 2 helper functions and removed the 2 other clunky ones that were too situation-specific:

- 5d2bd88: Implements `genericXDelCommand` for parsing across `XDEL`, `XDELEX`, and `XDELEX`
    - Named slightly different than how @zuiderkwast suggested to align with similarly named parsers for sorted set, hash, and kv commands
    - I stopped at parsing the count of ID's (and not actually parsing the IDs themselves into an array), becase of 2 reasons:
        1. Some commands need to reply (with varying response shapes) in between the initial parsing step and the individual IDs array parsing step
        2. Keeping the allocations in each function makes the memory management simpler, so all allocations are in the same spot in the commands and you don't have to think about whether the helper allocated or not.
- b814bcf: Adds `streamDeletePELEntry` centralizing 5 call sites across `XACK`, `XACKDEL` and `XDELEX`
- 98c4c30: Remove clunky XDELEX/XACKDEL helpers (`streamTrackFirstEntryAndPropagate` and `streamDeleteItemAndTrackFirstLast`)

I'm looking for other opportunities for deduplication and not seeing anything jump out. But if there's something I'm missing, happy to add. And if `streamDeletePELEntry` is also too situation specific or not widely shared enough, I'm happy to revert/adjust. Same for the balance struck for allocation / ID array parsing in `xdelGenericCommand`.

Additionally, `clang-format` is failing in CI due to an issue with packages. I can attest that (at least locally for me), `clang-format` is passing. I also edited my `.vimrc` to avoid noise like 71669bb & e9149a5.

**Comment by nickiaq** [comment]
One last folllowup here (37e6690), adding the fix to `XACKDEL` for the edge case called out by coderabbit earlier for `XDELEX`.

> I'm looking for other opportunities for deduplication and not seeing anything jump out.

I guess below is an example, where like with `streamTrackFirstEntryAndPropagate` and `streamDeleteItemAndTrackFirstLast`, there are a few snippets that are the same, but not sure if these are good candidates for helpers, as they are pretty specific to certain points in the logic as you mentioned.

```
// Appears in both XACKDEL & XDELEX

/* ACKED: check the PEL before consulting cg->last_id:
 * XGROUP SETID can move last_id backward below IDs that
 * are still pending (or were pending and later acked),
 * so last_id alone cannot prove this group never claimed
 * the message. */
unsigned char buf[sizeof(streamID)];
streamEncodeID(buf, id);
void *result;
if (raxFind(cg->pel, buf, sizeof(buf), &result)) {
    /* Still pending in this group, cannot delete. */
    resps[j] = 2;
} else if (exists[j] &&
           streamCompareID(id, &cg->last_id) > 0) {
    /* Message exists and may still be delivered to this
     * group, so block deletion (same as XACKDEL). Entries
     * that no longer exist can't be re-delivered. */
    resps[j] = 2;
}
```

**Review by zuiderkwast — COMMENTED** [review]
Hm... without code duplication, it's supposed to be fewer lines of code, but I see more lines. That doesn't seem right...

After the last change, I see +699 -20 for t_stream.c. In the individual PRs for XDELEX and XACKEL before the last commits, without any helper functions, I see +206 and +279 lines respectively.

If you're using an AI agent for this, I can understand, because they are good at adding new code, but very bad at fitting new functionality into existing code in a clean way. That needs a lot of human guidance, or just doing it manually.

**Comment by zuiderkwast** [comment]
I experimented with refactoring to reduce the code size locally. I pushed two commits to your branch. If you don't like them, you can remove them. 😆

My idea "XDEL is just XDELEX with KEEPREF" is almost correct, but apart from the syntax difference, the return value also differs: scalar vs array. Coupling "has pelmode arg" with the return value type didn't seem right, so I used an enum for which command it is instead of the two boolean arguments.

My second idea that "XDELEX is XACKDEL without group" is more or less right. We can implement it as group=NULL and use the same implementation for both and skip the first phase of XACKDEL if group is null.

With these two ideas, implementing all three commands using a single function with just some conditional code blocks reduces the size significantly, from +699/-20 to +380/-37 for `t_stream.c`.

When these commands are implemented using a single generic wrapper, there's no need to break out functions like streamParseStrictIDsOrReply – it'd just have a single caller, so we can inline it. Breaking out to a helper function doesn't reduce code duplication in that case. Saved a few more lines this way. It's easy to move it to a function if another command would need it in the future.

File: src/commands/xdelex.json [hunk 9]
```diff
@@ -0,0 +1,93 @@
+{
+    "XDELEX": {
+        "summary": "Delete stream message(s) with extended options",
+        "complexity": "O(1)",
```
Comment (hpatro) [inline src/commands/xdelex.json]:
```suggestion
        "complexity": "O(1) for each single item to delete in the stream, regardless of the stream size",
```

File: src/commands/xackdel.json [hunk 10]
```diff
@@ -0,0 +1,97 @@
+{
+    "XACKDEL": {
+        "summary": "Acknowledge and (if possible) delete stream message(s).",
+        "complexity": "O(1)",
```
Comment (hpatro) [inline src/commands/xackdel.json]:
```suggestion
        "complexity": "O(1) for each single item to delete in the stream, regardless of the stream size",
```

File: src/server.c [hunk 11]
```diff
@@ -7492,8 +7494,8 @@ void dismissMemoryInChild(void) {
     /* madvise(MADV_DONTNEED) may not work if Transparent Huge Pages is enabled. */
     if (server.thp_enabled) return;
 
-        /* Currently we use zmadvise_dontneed only when we use jemalloc with Linux.
-         * so we avoid these pointless loops when they're not going to do anything. */
+/* Currently we use zmadvise_dontneed only when we use jemalloc with Linux.
```
Comment (hpatro) [inline src/server.c]: Revert unless intended to.

File: src/t_stream.c [hunk 12]
```diff
@@ -3551,48 +3602,294 @@ void xautoclaimCommand(client *c) {
...
+     * it in Phase 2. XDELEX has no target group so this phase is skipped. */
+    if (group) {
+        for (size_t j = 0; j < id_count; j++) {
+            if (streamDeletePELEntry(group->pel, &ids[j])) {
+                acked++;
+                acked_flags[j] = 1;
+                /* resps[j] stays 1: eligible for deletion. */
+            } else {
+                resps[j] = -1;
+            }
```
Comment (hpatro) [inline src/t_stream.c:3794]: Are we conflating two different kind of behaviour into one i.e. missing entry and missing PEL entry? we are returning -1 for both the cases. Would it be required for client to distinguish between these states? What does Redis do?

**Review by hpatro — COMMENTED** [review]
@zuiderkwast overall structure looks good and compact. Didn't thoroughly check each of the conditions.

@nickiaq would you be able to add a test with other-server mechanism to verify the compatibility with an older version of Valkey.

**Comment by hpatro** [comment]
@nickiaq could you also update the top level comment to describe the feature. It will be part of the commit message. Thanks.

File: src/t_stream.c (hunk 12 above)
Comment (zuiderkwast) [reply src/t_stream.c:3794]: According to Redis XDELEX and XACKDEL docs:

> * Array reply: -1 for each requested ID when the given key does not exist.
> * Array reply: For each ID:
>   * Integer reply: -1 if no such ID exists in the provided stream key.
>   * ...

For existing entry without a PEL entry, it doesn't say... Missing PEL entry means that it's already been acked by some/every group?

Comment (nickiaq) [reply src/t_stream.c:3794]: Yes, this was surprising behavior for me too. This is what Redis does as well, for example tested below in Redis 8.10:

```
# Setup stream w/ 2 consumers & 2 messages; both consumers w/ last delivered id at message 1
> DEL mystream
(integer) 1
> XADD mystream 1 msg hello
"1-0"
> XGROUP CREATE mystream grp1 1
OK
> XGROUP CREATE mystream grp2 1
OK
> XADD mystream 2 msg2 hello2
"2-0"

# Read message 2 into group1 PEL; not in group2 PEL
> XREADGROUP GROUP grp1 cnsmr COUNT 1 STREAMS mystream >
1) 1) "testxadstream"
   2) 1) 1) "2-0"
         2) 1) "msg2"
            2) "hello2"
> XPENDING mystream grp2
1) (integer) 0
2) (nil)
3) (nil)
4) (nil)

#  Attempting to ACK a message not in group2's PEL results in -1
> XACKDEL mystream grp2 ACKED IDS 1 2-0
1) (integer) -1
```

I think the idea is that the `-1` status for an ID is also covering when the message wasn't acked as well.

File: src/server.c (hunk 11 above)
Comment (nickiaq) [reply src/server.c]: Reverted in eb2f13e.

**Comment by zuiderkwast** [comment]
> add a test with other-server mechanism to verify the compatibility with an older version of Valkey.

@hpatro means in tests/integration/cross-version-replication.tcl - use XDELEX or XACKDEL on the primary, verify that an old replica can replicate it (propagated as XDEL and XACK).

File: src/t_stream.c (hunk 12 above)
Comment (hpatro) [reply src/t_stream.c:3794]: I think so. Maybe for compatibility we should run their server and determine the return value (avoid looking at the code).

Comment (nickiaq) [reply src/t_stream.c:3794]: Sorry if I was unclear, the above comment is running Redis OSS 8.10 (without looking at the code).

```yaml
# docker-compose.yml
services:
  redis:
    image: 'redis:8.10-alpine'
    ports:
      - '127.0.0.1:6379:6379/tcp'
```

```
# test.txt
DEL mystream
XADD mystream 1 msg hello
XGROUP CREATE mystream grp1 1
XGROUP CREATE mystream grp2 1
XADD mystream 2 msg2 hello2

XREADGROUP GROUP grp1 cnsmr COUNT 1 STREAMS mystream >
XPENDING mystream grp2

XACKDEL mystream grp2 ACKED IDS 1 2-0
```

```bash
$ docker-compose up -d
$ valkey-cli < test.txt
(integer) 1
"1-0"
OK
OK
"2-0"
1) 1) "mystream"
   2) 1) 1) "2-0"
         2) 1) "msg2"
            2) "hello2"
1) (integer) 0
2) (nil)
3) (nil)
4) (nil)
1) (integer) -1  # <- response where message 2-0 isn't in grp2 PEL
```

**Comment by zuiderkwast** [comment]
Regarding cross-version-replication testing, there is already tests to verify what gets propagated. According to my AI tools:

----

Propagation is tested — there are 6 tests in stream-cgroups.tcl that verify the rewritten commands on replicas:
  
  - XACKDEL replication: ack-only propagates PEL removal to replica
  - XACKDEL replication: ACKED mode deletion propagates stream removal to replica
  - XACKDEL replication: DELREF clears other groups' PELs on replica
  - XDELEX replication: KEEPREF deletes stream entry but keeps dangling PEL ref on replica
  - XDELEX replication: DELREF deletes stream entry and clears PEL on replica
  - XDELEX replication: ACKED skips pending entries, deletes only after all groups ack
  
  And 6 more tests verify the shape of what gets propagated (checking that XACKDEL/XDELEX never appear in the propagation log, only XACK/XDEL):
  
  - XACKDEL ack-only propagates XACK but never XACKDEL or XDEL
  - XACKDEL KEEPREF propagates XACK + XDEL but never XACKDEL
  - XACKDEL DELREF propagates per-group XACK + XDEL but never XACKDEL
  - XDELEX KEEPREF propagates XDEL only but never XDELEX
  - XDELEX DELREF propagates XDEL + XACK but never XDELEX
  - XDELEX ACKED with nothing deleted propagates nothing

----

So I guess this is enough. @hpatro WDYT?

**Comment by nickiaq** [comment]
`tests/integration/cross-version-replication.tcl` is pretty cool. I wasn't aware of that spot to test this circumstance specifically.

Like @zuiderkwast mentioned, I had tests setup in `tests/unit/type/stream-cgroups.tcl` because that's where I saw `XCLAIM` tests covering a similar setup, ex. `Replication tests of XCLAIM with deleted entries`.

If `tests/integration/cross-version-replication.tcl` is a better spot, happy to move more cases there. cffdf2ee8 sets up a single test for `XACKDEL` and `XDELEX` using `DELREF` for both because that should exercise both the `XACK` and `XDEL` downgraded propagation portions.

**Comment by zuiderkwast** [comment]
> happy to move more cases there

Yeah, if you want, you can move some, but we don't need to check everything in both places. Or we can keep it as-is. It's good you added some case, but the cross-version-replication requires some special setup so it doesn't run by default when someone just runs `make test` locally, so I guess we can keep the ones that verify the propagation log too.

**Comment by nickiaq** [comment]
> Or we can keep it as-is.

Ok, if the `stream-cgroups.tcl` tests run in a more typical setup, it seems like that's a safer spot to have an initial check on the effects. And, then the integration tests cover in a real-world server.

@hpatro I've updated the PR description, if that serves as a good commit message.

@zuiderkwast The compact version you posted is very elegant. I think I was too hesitant to mess with `XDEL` and wasn't seeing them as a tiered extensions of each other where `XACKDEL` is `XDELEX++` and `XDELEX` is `XDEL++`.

Still outstanding, I have to create an issue in [valkey-io/valkey-doc](https://github.com/valkey-io/valkey-doc) as requested in [valkey-io.github.io PR659](https://github.com/valkey-io/valkey-io.github.io/pull/659)

**Review by hpatro — APPROVED** [review]
Some cross version compatibility tests are good to have. LGTM overall.

**Review by zuiderkwast — APPROVED** [review]
LGTM, thanks!
